File: src/main/java/com/mercadez/service/UsuarioService.java
package com.mercadez.service;

import com.mercadez.dto.*;
import com.mercadez.exception.*;
import com.mercadez.model.Usuario;
import com.mercadez.repository.UsuarioRepository;
import com.mercadez.security.JwtService;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UsuarioService {

    private final UsuarioRepository repo;
    private final PasswordEncoder   encoder;
    private final JwtService        jwt;

    public UsuarioService(UsuarioRepository repo, PasswordEncoder encoder, JwtService jwt) {
        this.repo    = repo;
        this.encoder = encoder;
        this.jwt     = jwt;
    }

    @Transactional
    public UsuarioResponse cadastrar(CadastroUsuarioRequest req) {
        if (repo.existsByEmail(req.getEmail())) {
            throw new NegocioException("Email já cadastrado.");
        }
        if (req.getCpf() != null && !req.getCpf().isBlank() && repo.existsByCpf(req.getCpf())) {
            throw new NegocioException("CPF já cadastrado.");
        }

        Usuario u = Usuario.builder()
                .nome(req.getNome())
                .email(req.getEmail().toLowerCase())
                .senha(encoder.encode(req.getSenha()))
                .cpf(req.getCpf())
                .build();

        return UsuarioResponse.de(repo.save(u));
    }

    @Transactional(readOnly = true)
    public LoginResponse login(LoginRequest req) {
        Usuario u = repo.findByEmail(req.getEmail().toLowerCase())
                .orElseThrow(CredenciaisInvalidasException::new);

        if (!encoder.matches(req.getSenha(), u.getSenha())) {
            throw new CredenciaisInvalidasException();
        }

        String token = jwt.gerarToken(u.getId(), u.getEmail(), "CLIENTE");
        return new LoginResponse(token, "Bearer", u.getPerfil().name(), u.getId(), u.getNome(), u.getEmail());
    }

    @Transactional(readOnly = true)
    public UsuarioResponse buscarPorId(Long id) {
        return UsuarioResponse.de(
                repo.findById(id).orElseThrow(() -> new NaoEncontradoException("Usuário não encontrado."))
        );
    }
}

File: app/services/usuario_service.py
"""Equivalente Python de service/UsuarioService.java"""
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core import security
from app.exceptions import CredenciaisInvalidasException, NaoEncontradoException, NegocioException
from app.models.usuario import Usuario
from app.schemas.usuario import CadastroUsuarioRequest, LoginRequest, LoginResponse, UsuarioResponse


def cadastrar(db: Session, req: CadastroUsuarioRequest) -> UsuarioResponse:
    email = req.email.lower()

    if db.scalar(select(Usuario).where(Usuario.email == email)) is not None:
        raise NegocioException("Email ja cadastrado.")

    if req.cpf and req.cpf.strip():
        if db.scalar(select(Usuario).where(Usuario.cpf == req.cpf)) is not None:
            raise NegocioException("CPF ja cadastrado.")

    usuario = Usuario(
        nome=req.nome,
        email=email,
        senha=security.hash_senha(req.senha),
        cpf=req.cpf,
    )
    db.add(usuario)
    db.commit()
    db.refresh(usuario)
    return UsuarioResponse.de(usuario)


def login(db: Session, req: LoginRequest) -> LoginResponse:
    email = req.email.lower()
    usuario = db.scalar(select(Usuario).where(Usuario.email == email))

    if usuario is None or not security.verificar_senha(req.senha, usuario.senha):
        raise CredenciaisInvalidasException()

    token = security.gerar_token(usuario.id, usuario.email, usuario.perfil.value)
    return LoginResponse(
        token=token,
        tipo="Bearer",
        perfil=usuario.perfil.value,
        id=usuario.id,
        nome=usuario.nome,
        email=usuario.email,
    )


def buscar_por_id(db: Session, usuario_id: int) -> UsuarioResponse:
    usuario = db.get(Usuario, usuario_id)
    if usuario is None:
        raise NaoEncontradoException("Usuario nao encontrado.")
    return UsuarioResponse.de(usuario)
